File: ios/FreakCore/Sources/FreakCore/Analysis.swift
// Analysis.swift — pure content-analysis functions over SlotRecords from
// anywhere (snapshot, backup). Port of analysis.py.
//
// Emptiness is a content judgement: the MicroFreak ships every unused slot
// as a factory Init preset with the name "Init", so a blank name never
// happens on a stock device and the string "Init" is never matched. Unknown
// disqualifies its own rule only: sha256 nil (content unread) can never
// satisfy the duplicate rule, and name nil (the name read FAILED — a
// swallowed timeout in snapshot — not a blank slot) can never satisfy the
// blank-name rule. The rules stay independent: a name-read-failed slot
// whose blob IS mass-duplicated is still expendable.

import Foundation

public enum Analysis {
    /// How many slots hold each blob hash; records without a hash are skipped.
    public static func shaCensus(_ records: [SlotRecord]) -> [String: Int] {
        var counts: [String: Int] = [:]
        for r in records {
            if let sha = r.sha256 {
                counts[sha, default: 0] += 1
            }
        }
        return counts
    }

    /// Expendable = successfully-read blank/whitespace name, OR sha256
    /// occurring >= threshold times among `records` (3, not 2, so a user's
    /// own single duplicated preset is never chosen). NEVER a name == "Init"
    /// match.
    public static func findExpendable(_ records: [SlotRecord],
                                      threshold: Int = Wire.duplicateThreshold) -> Set<Int> {
        let counts = shaCensus(records)
        var out: Set<Int> = []
        for r in records {
            guard let sha = r.sha256 else {
                continue
            }
            if let name = r.name,
               name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                out.insert(r.slot)
            } else if counts[sha, default: 0] >= threshold {
                out.insert(r.slot)
            }
        }
        return out
    }

    /// The safest slot to write into: the highest expendable slot >=
    /// preferFrom, else the highest expendable overall, else nil (the caller
    /// asks the human). Exactly the proven phase-0
    /// mfcap.verify.pick_scratch_slot semantics.
    public static func pickScratchSlot(_ records: [SlotRecord],
                                      preferFrom: Int = 500,
                                      exclude: Set<Int> = []) -> Int? {
        let expendable = findExpendable(records)
        for floor in [preferFrom, 0] {
            let picks = records.map(\.slot).filter {
                $0 >= floor && !exclude.contains($0) && expendable.contains($0)
            }
            if let best = picks.max() {
                return best
            }
        }
        return nil
    }
}

File: microfreak/analysis.py
"""Pure content-analysis functions. Accept SlotRecords from snapshots,
backups, anywhere.

Emptiness is a content judgement: the MicroFreak ships every unused slot as
a factory Init preset with the name "Init", so a blank name never happens on
a stock device and the string "Init" is never matched. A slot is expendable
when its exact bytes occur at least DUPLICATE_THRESHOLD times (3, not 2, so
a user's own single duplicated preset is never chosen), or its
successfully-read name is blank/whitespace-only (the phase-0 scratch rule).
Unknown disqualifies its own rule: a record whose sha256 is None (content
unread) can never satisfy the duplicate rule, and a record whose name is
None — the name READ FAILED (a swallowed timeout in MicroFreak.snapshot),
not a blank slot — can never satisfy the blank-name rule. The rules stay
independent: a name-read-failed slot whose blob IS mass-duplicated is still
expendable, because the content judgement doesn't need the name.
"""
from __future__ import annotations

from typing import Collection, Dict, Iterable, Optional, Set

from .model import SlotRecord
from .protocol import DUPLICATE_THRESHOLD


def sha_census(records: Iterable[SlotRecord]) -> Dict[str, int]:
    """How many slots hold each blob hash. Records without a hash are
    skipped."""
    counts: Dict[str, int] = {}
    for r in records:
        if r.sha256 is not None:
            counts[r.sha256] = counts.get(r.sha256, 0) + 1
    return counts


def find_expendable(records: Iterable[SlotRecord], *,
                    threshold: int = DUPLICATE_THRESHOLD) -> Set[int]:
    """Slots whose content is expendable: a successfully-read blank name, OR
    sha256 occurring >= threshold times. Never a name == "Init" string
    match. Unknown is never expendable: sha256 None (content unread) and
    name None (name read FAILED — MicroFreak.snapshot records None after a
    swallowed timeout/mismatch) both disqualify the respective rule, so a
    unique user preset with one transient name-read failure is never
    classified overwritable."""
    records = list(records)
    counts = sha_census(records)
    out: Set[int] = set()
    for r in records:
        if r.sha256 is None:
            continue
        if r.name is not None and not r.name.strip():
            out.add(r.slot)
        elif counts[r.sha256] >= threshold:
            out.add(r.slot)
    return out


def pick_scratch_slot(records: Iterable[SlotRecord], *, prefer_from: int = 500,
                      exclude: Collection[int] = ()) -> Optional[int]:
    """The safest slot to write into: the highest-numbered expendable slot
    >= prefer_from, else the highest expendable slot overall; None if
    nothing qualifies (the caller asks the human). Preserves the proven
    mfcap.verify.pick_scratch_slot semantics exactly."""
    records = list(records)
    expendable = find_expendable(records)
    excluded = set(exclude)
    for floor in (prefer_from, 0):
        picks = [r.slot for r in records
                 if r.slot >= floor and r.slot not in excluded
                 and r.slot in expendable]
        if picks:
            return max(picks)
    return None
